src/TraceStream.cc

```diff
@@ -1073,35 +1074,110 @@ void TraceWriter::close() {
   }
 }
 
-static string make_trace_dir(const string& exe_path) {
-  ensure_default_rr_trace_dir();
-
-  // Find a unique trace directory name.
-  int nonce = 0;
-  int ret;
-  string dir;
-  do {
-    stringstream ss;
-    ss << trace_save_dir() << "/" << basename(exe_path.c_str()) << "-"
-       << nonce++;
-    dir = ss.str();
-    ret = mkdir(dir.c_str(), S_IRWXU | S_IRWXG);
-  } while (ret && EEXIST == errno);
-
-  if (ret) {
-    FATAL() << "Unable to create trace directory `" << dir << "'";
-  }
+static int remove_directory(const char* dir) {
+	int ret = 0;
+	DIR* d = opendir(dir);
+	if (d == NULL) {
+		return -1;
+	}
+	struct dirent* next_file;
+	char filepath[256];
+
+	while ((next_file = readdir(d)) != NULL) {
+		if (0 == strcmp(next_file->d_name, ".") || 0 == strcmp(next_file->d_name, "..")) {
+			continue;
+		}
+		sprintf(filepath, "%s/%s", dir, next_file->d_name);
+		ret = remove(filepath);
+		if (ret < 0) {
+			FATAL() << "Unable to remove file`" << filepath << "'";
+		}
+	}
+	ret = closedir(d);
+	if (ret < 0) {
+		FATAL() << "Unable to close directory`" << dir << "'";
+	}
+	ret = rmdir(dir);
+	if (ret < 0) {
+		FATAL() << "Unable to remove directory`" << dir << "'";
+	}
+
+	return ret;
+}
 
-  return dir;
+static string make_trace_dir(const string& exe_path, const string& output_trace_dir) {
+
+	// save trace dir in given trace dir with option -o
+	if (output_trace_dir != "") {
+		DIR* directory = opendir(output_trace_dir.c_str());
+		if (directory) {
+			// directory exist, remove all the files in directory and delete it
+			int res = remove_directory(output_trace_dir.c_str());
```

Comment:
`remove_directory` doesn't delete directories with subdirectories inside them. Maybe that's a good thing, but really I'd prefer it if you didn't delete the directory at all; just error out if the directory exists and is non-empty.